addons/nuxeo-core-binarymanager-cloud/nuxeo-core-binarymanager-gcp/src/main/java/org/nuxeo/ecm/core/storage/gcp/GoogleStorageBinaryManager.java

```diff
@@ -0,0 +1,241 @@
+                log.debug(String.format("Creating a new bucket %s", bucketName));
+                return storage.create(BucketInfo.of(bucketName));
+            }
+            return bucket;
+        } catch (StorageException e) {
+            throw new NuxeoException(e);
+        }
+    }
+
+    /**
+     * Deletes a bucket (and all its blobs) with the given {@code bucketName}.
+     *
+     * @return boolean if bucket has been deleted or not.
+     */
+    public boolean deleteBucket(String bucketName) {
+        try {
+            Bucket bucket = storage.get(bucketName);
+            for (Blob blob : storage.list(bucketName).iterateAll()) {
+                blob.delete();
+            }
+            return bucket.exists() && storage.delete(bucketName);
+        } catch (StorageException e) {
+            throw new NuxeoException(e);
+        }
+    }
+
+    public Bucket getBucket() {
+        return bucket;
+    }
+
+    @Override
+    protected FileStorage getFileStorage() {
+        return new GCPFileStorage();
+    }
+
+    public class GCPFileStorage implements FileStorage {
+
+        @Override
+        public void storeFile(String key, File file) {
+            try {
+                bucket.create(key, new FileInputStream(file));
+            } catch (StorageException | IOException e) {
+                throw new NuxeoException(e);
+            }
+        }
+
+        @Override
+        public boolean fetchFile(String key, File file) {
+            try {
+                Blob blob = bucket.get(key);
+                if (blob != null) {
+                    blob.downloadTo(file.toPath());
+                    return true;
+                }
+                return false;
+            } catch (StorageException e) {
+                throw new NuxeoException(e);
+            }
+        }
+    }
+
+    @Override
+    protected String getSystemPropertyPrefix() {
+        return SYSTEM_PROPERTY_PREFIX;
+    }
+
+    /**
+     * Garbage collector for GCP binaries that stores the marked (in use) binaries in memory.
+     */
+    public static class GoogleStorageBinaryGarbageCollector
+            extends AbstractBinaryGarbageCollector<GoogleStorageBinaryManager> {
+
+        protected GoogleStorageBinaryGarbageCollector(GoogleStorageBinaryManager binaryManager) {
+            super(binaryManager);
+        }
+
+        @Override
+        public String getId() {
+            return "gcs:" + binaryManager.bucketName;
+        }
+
+        protected String getDigest(Blob blob) {
+            return blob.getBlobId().getName();
+        }
+
+        @Override
+        public Set<String> getUnmarkedBlobs() {
+            try {
+                Set<String> unmarked = new HashSet<>();
+                Page<Blob> blobs = binaryManager.getBucket().list(BlobListOption.fields(BlobField.ID, BlobField.SIZE));
+                do {
+                    for (Blob blob : blobs.iterateAll()) {
+                        if (marked.contains(getDigest(blob))) {
```

Comment:
Please add a safety check like for S3 that it's really a MD5 digest, to avoid deleting unwanted files.